KCORE-180: Include features added for 5.5.x (#537)
confluentinc/examples: multiregion/README.md

### Broker
* `broker.rack`: identifies the location of the broker. For the demo, it represents a region, either `east` or `west`
* `replica.selector.class=org.apache.kafka.common.replica.RackAwareReplicaSelector`: allows clients to read from followers (in contrast, clients are typically only allowed to read from leaders)

> * `confluent.log.placement.constraints={"version":1,"replicas":[{"count":2,"constraints":{"rack":"west"}}],"observers":[{"count":2,"constraints":{"rack":"east"}}]}`: This sets the default replica placement constraint configuration to use when creating a topic.

Reviewer: ```suggestion
* `confluent.log.placement.constraints`: sets the default replica placement constraint configuration for newly created topics.
```

The replica placement policy file is defined with the argument `--replica-placement <path-to-replica-placement-policy-json>` mentioned earlier (these files are in the [config](config/) directory).
Each placement also has a minimum `count` associated with it that allows users to guarantee a certain spread of replicas throughout the cluster.

| Topic name           | Leader  | Followers (sync replicas) | Observers (async replicas) | ISR list  |
|----------------------|---------|---------------------------|----------------------------|-----------|
| single-region        | 1x west | 1x west                   | n/a                        | {1,2}     |
| multi-region-sync    | 1x west | 1x west, 2x east          | n/a                        | {1,2,3,4} |
| multi-region-async   | 1x west | 1x west                   | 2x east                    | {1,2}     |

> | multi-region-default | 1x west | 1x west                   | 2x east                    | {1,2}     |

Reviewer: From this table, `multi-region-async` looks identical to `multi-region-default`.  Users will be confused as to what the difference is.  Can you please add a column that represents the difference?

Topic: multi-region-async	PartitionCount: 1	ReplicationFactor: 4	Configs: min.insync.replicas=1,confluent.placement.constraints={"version":1,"replicas":[{"count":2,"constraints":{"rack":"west"}}],"observers":[{"count":2,"constraints":{"rack":"east"}}]}
	Topic: multi-region-async	Partition: 0	Leader: none	Replicas: 2,1,3,4	Isr: 1	Offline: 2,1	Observers: 3,4

==> Describe topic multi-region-default

Topic: multi-region-default	PartitionCount: 1	ReplicationFactor: 4	Configs: min.insync.replicas=1,confluent.placement.constraints={"version":1,"replicas":[{"count":2,"constraints":{"rack":"west"}}],"observers":[{"count":2,"constraints":{"rack":"east"}}]}
	Topic: multi-region-default	Partition: 0	Leader: none	Replicas: 2,1,3,4	Isr: 1	Offline: 2,1	Observers: 3,4
```

Observations:

* In the first case, the topic `single-region` has no leader, because it had only two replicas in the ISR, both of which were in the `west` region and are now down.
* In the second case, the topic `multi-region-sync` automatically elected a new leader in `east` (e.g. replica 3 in the above output).  Clients can failover to those replicas in the east region.

> * In the third case, the topics `multi-region-async` and `multi-region-default` also have no leader, because they had only two replicas in the ISR, both of which were in the `west` region and are now down.  The observers in the `east` region are not eligible to become leaders automatically because they were not in the ISR.

Reviewer: ```suggestion
* In the last two cases, the topics `multi-region-async` and `multi-region-default` also have no leader, because they had only two replicas in the ISR, both of which were in the `west` region and are now down.  The observers in the `east` region are not eligible to become leaders automatically because they were not in the ISR.
```

Topic: multi-region-async	PartitionCount: 1	ReplicationFactor: 4	Configs: min.insync.replicas=1,confluent.placement.constraints={"version":1,"replicas":[{"count":2,"constraints":{"rack":"west"}}],"observers":[{"count":2,"constraints":{"rack":"east"}}]}
	Topic: multi-region-async	Partition: 0	Leader: 3	Replicas: 2,1,3,4	Isr: 3,4	Offline: 2,1	Observers: 3,4

==> Describe topic multi-region-default

Topic: multi-region-default	PartitionCount: 1	ReplicationFactor: 4	Configs: min.insync.replicas=1,confluent.placement.constraints={"version":1,"replicas":[{"count":2,"constraints":{"rack":"west"}}],"observers":[{"count":2,"constraints":{"rack":"east"}}]}
	Topic: multi-region-default	Partition: 0	Leader: 3	Replicas: 2,1,3,4	Isr: 3,4	Offline: 2,1	Observers: 3,4
```

Observations for topics `multi-region-async` and `multi-region-default`:

> * It has a leader again (e.g. replica 3 in the above output)

Reviewer: ```suggestion
* They have leaders again (e.g. replica 3 in the above output)
```

* The observers are now in the ISR list (e.g. replicas 3,4 in the above output)

### Permanent Failover

> At this point in the example, if the brokers `broker-west-1` and `broker-west-2` come back online, then by default the leaders for the topics `multi-region-async` and `multi-region-default` will be elected back to a replica with a rack in `west` (i.e., replica 1 or 2). To change this behavior we can change the topic placement constraints configuration and replica assignment of the affected topic. Here we are going to change the topic placement constraints configuration and replica assignment for `multi-region-default`.

Reviewer: This paragraph is a bit hard to understand. Here is a rephrase, please let me know if I understood it correctly.

> At this point in the example, if the brokers in the `west` region come back online, then by default the leaders for the topics `multi-region-async` and `multi-region-default` will automatically be elected back to a replica in `west` (i.e., replica 1 or 2). This may be desirable in some circumstances, but if you don't want them to automatically failback, change the topic placement constraints configuration and replica assignment. In the next step, change the topic placement constraints configuration and replica assignment for `multi-region-default`.

Author: Yes, that is correct. Done.

Topic: multi-region-async	PartitionCount: 1	ReplicationFactor: 4	Configs: min.insync.replicas=1,confluent.placement.constraints={"version":1,"replicas":[{"count":2,"constraints":{"rack":"west"}}],"observers":[{"count":2,"constraints":{"rack":"east"}}]}
	Topic: multi-region-async	Partition: 0	Leader: 3	Replicas: 2,1,3,4	Isr: 3,4	Offline: 2,1	Observers: 3,4

==> Describe topic multi-region-default

Topic: multi-region-default	PartitionCount: 1	ReplicationFactor: 4	Configs: min.insync.replicas=1,confluent.placement.constraints={"version":1,"replicas":[{"count":2,"constraints":{"rack":"west"}}],"observers":[{"count":2,"constraints":{"rack":"east"}}]}
	Topic: multi-region-default	Partition: 0	Leader: 3	Replicas: 2,1,3,4	Isr: 3,4	Offline: 2,1	Observers: 3,4
```

Observations for topics `multi-region-async` and `multi-region-default`:

* It has a leader again (e.g. replica 3 in the above output)
* The observers are now in the ISR list (e.g. replicas 3,4 in the above output)

### Permanent Failover

> At this point in the example, if the brokers `broker-west-1` and `broker-west-2` come back online then by default the leaders for the topics `multi-region-async` and `multi-region-default` will be elected back to a replica with a rack of `west` (replica 1 or 2). To change this fallback behavior we can change the configuration of assignment of the affected topic. Here we are going to perform a permanent fallback to the east region for the topic `multi-region-default`.

Reviewer: ```suggestion
At this point in the example, if the brokers `broker-west-1` and `broker-west-2` come back online, then by default the leaders for the topics `multi-region-async` and `multi-region-default` will be elected back to a replica with a rack in `west` (i.e., replica 1 or 2). To change this fallback behavior we can change the configuration of assignment of the affected topic. Here we are going to perform a permanent fallback to the east region for the topic `multi-region-default`.
```

Reviewer: What does "configuration of assignment of the affected topic" mean?

Reviewer: IIUC, `fallback` is not the right term, I think it should be `failover`?

Author: Okay. I don't think `failover` is accurate either since it is not a failure. How about:

> At this point in the example, if the brokers `broker-west-1` and `broker-west-2` come back online, then by default the leaders for the topics `multi-region-async` and `multi-region-default` will be elected back to a replica with a rack in `west` (i.e., replica 1 or 2). To change this behavior we can change the topic placement constraints configuration and replica assignment of the affected topic. Here we are going to change the topic placement constraints configuration and replica assignment for `multi-region-default`.
